rfcs/rfc0017.md

```diff
@@ -0,0 +1,149 @@
+accurately serialize values.
+
+## Rationale
+
+While the flags on scalars can be checked via the B module, using it requires
+an understanding of perl internals, and it is slow and uses a fairly large
+amount of memory. Being able to interpret the flags to know string vs number
+is also not entirely obvious.
+
+The functions are intentionally using longer less convenient names, with the
+hope that this will discourage their use for type checks within perl. Inside
+perl, 1 and "1" are meant to be treated equivalently. It is only when
+transferring data outside perl that these are meant to be used.
+
+`created_as` is meant to only tell how a value was created, without implying
+anything further about what the value "is". `created` in this case means any
+modification of the value, as any modification of a value requires creating
+the new, modified value.
+
+## Specification
+
+    created_as_string VALUE
+
+`created_as_string` takes a single scalar argument.
+
+It returns a boolean value representing if the value was originally created as
+a string. It will return false for references, numbers, booleans, and undef.
+Internally, this would be `SvPOK(sv) && !SvIsBOOL(sv)`.
+
+    created_as_number VALUE
+
+`created_as_number` takes a single scalar argument.
+
+It returns a boolean value representing if the value was originally created as
+a number. It will return false for references, strings, booleans, and undef.
+Internally, this would be `SvNIOK(sv) && !SvPOK(sv) && !SvIsBOOL(sv)`.
+
+## Backwards Compatibility
+
+There should be no significant backwards compatibility concerns.
+`created_as_string` and `created_as_number` will be imported only when
+requested. Static analysis tools may need to be updated.
+
+As these functions rely on the changes to how SV flags get set, it isn't
+possible to accurately implement these on older perls. Existing serializers
+use heuristics that will usually give accurate answers. One of these heuristics
+could be used to provide a compatible polyfill.
+
+## Security Implications
+
+Nothing specific predicted.
+
+## Examples
+
+  use builtin qw(created_as_number created_as_string);
+
+  my $value1 = "1";
+  my $string1 = created_as_string $value1; # true
+  my $number1 = created_as_number $value1; # false
+
+  my $value2 = 1;
+  my $string2 = created_as_string $value2; # false
+  my $number2 = created_as_number $value2; # true
+
+  my $value3 = "1";
+  my $used_as_number3 = 0+$value3;
+  my $string3 = created_as_string $value2; # true
+  my $number3 = created_as_number $value2; # false
+
+  my $value4 = 1;
+  my $used_as_string4 = "$value4";
+  my $string4 = created_as_string $value2; # false
+  my $number4 = created_as_number $value2; # true
+
+## Prototype Implementation
+
+None.
+
+## Future Scope
+
+It is possible that in the future, the NOK or IOK flags will only be set for
+values created as numbers. That would mean the implementation of these
+functions may change, but should not impact the results returned.
```

Comment:
One slightly awkward upshot is that in the current implementation, values that start off as strings can still gain the NOK or IOK flags. Because of that, it means we can't distinguish created-as-strings from genuine dualvars (either core variables like `$!` or the result of `Scalar::Util::dualvar`).

There's not much we can do about that at present, but it may be something we want to take note of when documenting how `created_as_string` works at the moment. A future version may not be confused by dualvars, but currently both of these will be true:

```perl
created_as_string $!;
created_as_string dualvar(123, "four-five-six");
```